pkg/sql/conn_executor_test.go

```diff
@@ -970,6 +973,182 @@ func TestShowLastQueryStatisticsUnknown(t *testing.T) {
 	require.False(t, y.Valid)
 }
 
+// TestTransactionDeadline tests that the transaction deadline is set correctly:
+// - In a single-tenant environment, the transaction deadline should use the leased
+//   descriptor expiration.
+// - In a multi-tenant environment, the transaction deadline should use the
+//   min(sqlliveness.Session expiry, lease descriptor expiration) while
+//   setting the deadline.
+func TestTransactionDeadline(t *testing.T) {
+	defer leaktest.AfterTest(t)()
+	defer log.Scope(t).Close(t)
+
+	ctx := context.Background()
+	var txnDeadline hlc.Timestamp
+	var txnID string
+	var sqlLiveness slprovider.TestLiveness
+	timeSource := timeutil.NewTestTimeSource()
+	clock := hlc.NewClock(func() int64 {
+		return timeSource.Now().UnixNano()
+	}, base.DefaultMaxClockOffset)
+
+	// Set up a kvserverbase.ReplicaRequestFilter which will extract the deadline for the test transaction.
+	checkTransactionDeadlineFilter := func(_ context.Context, ba roachpb.BatchRequest) *roachpb.Error {
+		if ba.Txn == nil {
+			return nil
+		}
+		currentTxnID := ba.Txn.TxnMeta.ID.String()
+		if currentTxnID != txnID {
+			return nil
+		}
+
+		if args, ok := ba.GetArg(roachpb.EndTxn); ok {
+			et := args.(*roachpb.EndTxnRequest)
+			if et.Deadline == nil {
+				return nil
+			}
+			txnDeadline = *et.Deadline
+		}
+		return nil
+	}
+
+	// Use the BeforeStart transaction hook to inject a test SQLLiveness instance.
+	injectSQLLiveness := func(cfg *sql.ExecutorConfig) {
+		cfg.SQLLiveness = sqlLiveness
```

Comment:
this config struct is global so this feels pretty gross to me. What if instead you injected a much less general knob that allowed you to override the `sqlliveness.Session` and if it returns something avoids calling the `Session` method on the `sqlliveness.Instance`?